### DISCUSSION TITLE
Remember me feature error

### DISCUSSION THREAD

**Discussion by tipitjafar** [codeigniter4/shield#950] [Q&A] [answered]
Hello People,

First want to say thank you for this great framework. Awesome work. With that being said, want to ask if there is some issue with remember me feature, because for me isn't working well. If I check the remember me? input at login form, session registers ok and token lasts 30 days. But if i do not check the remember me? input, just simply gives me a login fail. If i do not check remember me, it is supposed to keep session till browser is closed, isnt that right? What i'm doing wrong?

**Comment by kenjis** [comment]
It is very strange. It is not supposed. I don't know why you get the strange results.

> But if i do not check the remember me? input, just simply gives me a login fail.

Of course it should  be successful without remember me cookie.

> If i do not check remember me, it is supposed to keep session till browser is closed, isnt that right?

No. The default session cookie lifetime is 7200 seconds.
See CodeIgniter's `Config\Session::$expiration`.

**Reply by tipitjafar** [reply]
Thanks you! You're right, session's cookie lifetime is set to 7200. I'll do some research and testing and see if I can get a more accurate view of the problem.

**Reply by tipitjafar** [reply]
Hello kenjis, mind me if i'm wrong but i've arrived to conclusion that the problem is that if you do not set remember me checkbox to true there is no mechanism that sets a session cookie (the login process returns successful anyways, i was wrong about that) and i'm affraid that the shield framework doesn't look for any other cookie than the one in `Config\Auth::sessionConfig['rememberCookieName']`

Here is the function that is not executed when remember me input is false, and no cookie session is set at all:

https://github.com/codeigniter4/shield/blob/368274b14af4aecda4542c17028ad91d100d5635/src/Authentication/Authenticators/Session.php#L765-L784

**Reply by kenjis** [reply]
Shield uses Session. It sets a session cookie.
https://github.com/codeigniter4/shield/blob/368274b14af4aecda4542c17028ad91d100d5635/src/Authentication/Authenticators/Session.php#L666

**Reply by tipitjafar** [reply]
Adding this modifications to shield/src/Authentication/Authenticators/Session.php problem es "solved" or patched

Got to add that if the cookie is name other than Config\Auth::sessionConfig['rememberCookieName'] i get the same not working result

Added code is placed below comment //THIS IS NEW
```php
private function issueRememberMeToken(): void
    {
        if ($this->shouldRemember && setting('Auth.sessionConfig')['allowRemembering']) {
            $this->rememberUser($this->user);

            // Reset so it doesn't mess up future calls.
            $this->shouldRemember = false;
        } elseif ($this->getRememberMeToken()) {
            $this->removeRememberCookie();

            // @TODO delete the token record.
        }
        //THIS IS NEW
        else {
          $this->registerUserCookie($this->user);
        }

        // We'll give a 20% chance to need to do a purge since we
        // don't need to purge THAT often, it's just a maintenance issue.
        // to keep the table from getting out of control.
        if (random_int(1, 100) <= 20) {
            $this->rememberModel->purgeOldRememberTokens();
        }
    }

    //THIS IS NEW
    protected function registerUserCookie(User $user): void
    {
        $selector  = bin2hex(random_bytes(12));
        $validator = bin2hex(random_bytes(20));
        $timestamp = Time::now()->getTimestamp() + setting('Session.expiration');
        $expires   = Time::createFromTimestamp($timestamp)->format('Y-m-d H:i:s');

        $rawToken = $selector . ':' . $validator;

        // Store it in the database.
        $this->rememberModel->rememberUser(
            $user,
            $selector,
            $this->hashValidator($validator),
            $expires
        ); 

        $response = service('response');
  
        // Save it to the user's browser in a cookie.
        // Create the cookie

        $response->setCookie(
            setting('Auth.sessionConfig')['rememberCookieName'],
            $rawToken,                                             // Value
            setting('Session.expiration'),      // # Seconds until it expires
            setting('Cookie.domain'),
            setting('Cookie.path'),
            setting('Cookie.prefix'),
            setting('Cookie.secure'),                          // Only send over HTTPS?
            true                                                  // Hide from Javascript?
        );
    }
```

**Reply by tipitjafar** [reply]
One more thing, the only code i have to protect the entire site is a default filter at config/Filters.php

```php
 public array $globals = [
        'before' => [
            // 'honeypot',
            // 'csrf',
            // 'invalidchars',
            'session' => ['except' => ['login*', 'register', 'auth/a/*']],
        ],
        'after' => [
            'toolbar',
            // 'honeypot',
            // 'secureheaders',
        ],
    ];
```

**Reply by kenjis** [reply]
You seems to go the wrong direction.
Shield uses CodeIgniter's Session
https://codeigniter4.github.io/CodeIgniter4/libraries/sessions.html
and it issues a session cookie by default.
So it is very odd if you don't get the session cookie.

**Reply by kenjis** [reply]
Did you customize the login process?

**Reply by tipitjafar** [reply]
> Did you customize the login process?

Much thank you for being so kind helping here. Nope, never customized a thing, should I? Just to not getting into any trouble just installed CI, installed shield, created an user via CLI, and created a simple controller and a filter to test how it goes. Nothing else.

Now you mentioned odd things, since i've stepped into CI4, on a previous project using version 4.2.7, sessions thing was a nightmare to get it working and fixed it adding a session_start() on public/index.php before $app->run() line, otherwise the ci_session value was changing on every call. That was odd

**Reply by kenjis** [reply, marked as answer]
It appears that there is something causing CI4 Session not to work in your environment.
Normally, just call `session()`, the session works fine, and you will get the session cookie.
Of course, hack like adding a `session_start()` is not needed.

Unfortunately, however, I do not know what is causing such a phenomenon.

**Reply by tipitjafar** [reply]
Dont worry pal, me neither. I've tried every workaround someone possibly could do, and nothing changed. Tried permissions, several servers, serveral php versions, several operating systems.. The only hack that worked was the session_start()... Anyways, i will be watching this thing on the future hoping to find a solution.

Thank you very much for your time.

**Reply by kenjis** [reply]
This is my test repository:
https://github.com/kenjis/ci4-shield-test
This works fine in my environment. I can register and login.
If it does not work, probably there is something wrong with your PHP.

**Reply by tipitjafar** [reply]
Much appreciated! Thank you!

**Reply by tipitjafar** [reply]
Hello kenjis, good news, but the problem is still unknown. Trying your test repository I've solved the problem changing the httpd service. Just running the PHP’s built-in server with "php spark serve" just did the trick. I cant believe i've never tried this workaround or suspected nothing about apache. Anyways, the problem was always apache, i've tryed several apache versions and distros and always the same error, cant believe i didn't see it. Now i'm migrating to nginx, wish me look.

Got to report this issue to codeigniter github, the problem is something with CI4 and default apache settings. Something around memory allocation or something like that ,because sessions works fine ouside codeigniter framework, but when you start stacking all the function calls, something starts to get missing. Maybe one day put me into debuging this

Again, thank you for your time. 

Cheers!

Mariano.
